fontforge/featurefile.c

```diff
@@ -6208,30 +6158,39 @@ return( type==ft_lookup_end || type == ft_feat_end ||
 	    type == ft_lookup_ref );
 }
 
+/* This function copies entries of a given lookup type into the
+ * lookup_next list of the first argument.  A given lookup is
+ * considered to end on encountering one of the types in
+ * fea_FeatItemEndsLookup or NULL, which is returned.
+ *
+ * When called with type==ot_undef the types are not checked
+ * and assumed to match, a condition verified by some of the
+ * variants of fea_ApplyLookupList.
+ *
+ * When called with another type the function adds elements of
+ * that type to lookup_next until either a list-terminating element
+ * or one of a different type is found.
+ */
 static struct feat_item *fea_SetLookupLink(struct feat_item *nested,
 	enum otlookup_type type) {
     struct feat_item *prev = NULL;
     enum otlookup_type found_type;
 
-    while ( nested!=NULL ) {
-	/* Stop when we find something which forces a new lookup */
-	if ( fea_FeatItemEndsLookup(nested->type) )
-    break;
-	if ( nested->ticked ) {
-	    nested = nested->next;
-    continue;
-	}
-	found_type = fea_LookupTypeFromItem(nested);
-	if ( type==ot_undef || found_type == ot_undef || found_type == type ) {
-	    if ( nested->type!=ft_ap || nested->u2.ap->type!=at_mark )
-		nested->ticked = true;		/* Marks might get used in more than one lookup */
-	    if ( prev!=NULL )
-		prev->lookup_next = nested;
-	    prev = nested;
+    while ( nested!=NULL && !fea_FeatItemEndsLookup(nested->type) ) {
+	if ( !nested->ticked ) {
+	    found_type = fea_LookupTypeFromItem(nested);
+	    if ( type==ot_undef || found_type == ot_undef || found_type == type ) {
+		if ( nested->type!=ft_ap || nested->u2.ap->type!=at_mark )
+		    nested->ticked = true; /* Marks might get used in more than one lookup */
+		if ( prev!=NULL )
+		    prev->lookup_next = nested;
+		prev = nested;
+	    } else if ( type!=ot_undef && found_type!=type )
```

Comment:
Is this (breaking on type mismatch) the only logical change in this function?

Reply:
Yes, although it's worth noting that that makes the return type rather different when not passing `ot_undef`. Before it would go through the whole list of lookups and return the first `feat_item` after. Now it returns intermediate values. As it happens that doesn't matter because the return value isn't used when the function is called this way.

Reply:
> As it happens that doesn't matter because the return value isn't used when the function is called this way.

In `fea_ApplyFeatureList` as you mention, for the case in question it doesn't matter, but I do see it would have an effect on the `ft_lookup_start` case. I'm still trying to understand this, but from what I get so far, we're now breaking up the lookup/definition when the type changes, vs. carrying on and applying that single substitution adjustment?

Reply:
Oh right, I see, it's called with `ot_undef` in that case

Reply:
This is all heavy spaghetti. I believe the other cases correspond to lookup-level directives rather than feature-level directives. So in those the types are supposed already be verified as uniform (that check is still there), and you'd get the same result here even if I changed the logic for `ot_undef`. But I didn't anyway.